### ISSUE TITLE
Use Proper Aria Roles on Alerts

### ISSUE THREAD

**Issue by srsgores** [wet-boew/wet-boew#6431]
I've noticed that all ``.alert`` elements fail to comply with the ARIA spec which recommends that the attribute ``role="alert"`` be added for accessibility.

This should be applied across the board.

**Comment by pjackson28** [comment]
The alert role is only intended to be used for ARIA live regions (which the alert role sets the live region to assertive). The <code>.alert</code> elements are not ARIA live regions by default (content is not expected to change or to suddenly appear). It is possible to make them ARIA live regions, which then an alert role could be added if assertive updates are needed (in other words immediately interrupting the user with the alert). By default all regions are set to <code>aria-live="off"<|code> but optionally it could be set to "polite" or "assertive" (through a role or the aria-live attribute).

So the alert role shouldn't be added by default, but that role or another role can be added where appropriate.

**Comment by srsgores** [comment]
@pjackson28, this is new information to me (though I now did see that you are correct in that ``aria-live`` must be off.  I still find it's not ideal to simply have a ``div`` element housing the alert content when clearly, the content *is* an alert.  When reviewing the [information on mdn](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role), it seems sensible to keep the role.

As you suggest, something like this may be better for assistive technologies:

```
<div class="alert alert-error" role="alert" aria-live="off">
    ...
</div>
```

And in the event that the alert becomes dismissible, that role should change to ``alert-dialog``.

**Comment by pjackson28** [comment]
I still don't think there should be a direct mapping between Bootstrap's alert class and the WAI-ARIA alert role. They may technically share the same name but they don't serve the same purpose. For instance Bootstrap's alert-info class could hardly be considered important and time-sensitive information like you would expect for an alert role. In that case it is a "For your information" and not an "Alert" so an alert role would misrepresent the purpose.

Also what purpose does an alert role serve if aria-live is off? In essence the alert role is just a group of presets for ARIA states and properties, the main of which are aria-live and aria-atomic. Should we continue to use a role if we need to change one of it's key traits?

**Comment by pjackson28** [comment]
There may be specific use cases where the alert role makes sense (like the form error example on MDN) but it shouldn't be for all cases. Most places (including on MDN) say to use alert "sparingly".

Maybe there should be guidance on when the alert role would be appropriate for an alert element. Something that would go in the styleguide. What use cases do you think the alert role would be appropriate for an alert element?

**Comment by melodyma** [comment]
Would an inline field error message be a candidate for the alert role or not?

**Comment by srsgores** [comment]
@melodyma, yes, if that message may be removed when the field(s) become validated; that's a perfect case for ``aria-live``.

**Comment by melodyma** [comment]
@srsgores Yes, the fields would be removed when the field(s) become validation.
Then are you meaning that you'd need to put both aria-live="on" and role="alert" on the same element or just role="alert"?

**Comment by srsgores** [comment]
@melodyma, no.  Like @pjackson28 said, by using ``role="alert"``, you are **implicitly** declaring ``aria-live`` on the element.

So putting ``role="alert"`` is enough.

**Comment by pjackson28** [comment]
Actually an inline form error message turned out to be a bad place for it, as we found out with the form validation plugin. We originally had it there but the end result was chatty to the point that the user couldn't do anything. One major problem with how screen readers and browsers deal with live regions is they don't check to see if the included content has changed, just that the aria-live region was updated. So the following happened:

1. Form was submitted generating errors. Because all the inline errors had the alert role, they were all immediately read out (very chatty and confusing if you have more than one error). On every subsequent keypress the errors would be read out in full again (since jQuery validation refreshed the error messages). This effectively prevented the user from doing anything because the user kept getting interrupted.

The way we dealt with this is to have a single hidden aria-live region that was only updated once the error message for the current field actually changed. This dramatically reduced the chattiness and made the form usable again (so only alerts when appropriate).

So the above example shows how careful we have to be about applying the alert role. It is still a valuable role, but we need to make sure that it's use doesn't do more harm than good.

**Comment by melodyma** [comment]
@pjackson28 Very insightful, thanks. 

I noticed that the inline form errors for wet-boew is presented post-form submission (http://wet-boew.github.io/v4.0-ci/demos/formvalid/formvalid-en.html). However, if the form fields were validated as you are filling out the form (pre-form submission, example here: http://parsleyjs.org/doc/examples/simple.html), I wonder if that would resolve the chatty screen reader problem you pointed out above.

**Comment by pjackson28** [comment]
Actually it does validation pre-submission as well. It does so after you type characters and the field loses focus. The problem though is if you return to that field and try to correct it you will hear the message over and over again on every keypress (since it refreshes on every keypress). Even worse, the interruption interrupts even the character being displayed so the field can never be corrected. So the alert role on an element that contains a jQuery validation inline error message is a bad combination.

**Comment by srsgores** [comment]
@pjackson28, ok then.  I have learned much from this issue.  I will close now.

**Comment by melodyma** [comment]
@pjackson28 Could you point out where I can see how you implemented the "single hidden aria-live region" you mentioned above? Someone needs to write a blog post on this! Thanks!

**Comment by pjackson28** [comment]
@melodyma The plugin creates the following hidden element:

```html
<div class="arialive wb-inv" aria-relevant="all" aria-live="polite"></div>
```

When there is an inline error message for the field (and it is changed from it's previous state) then the element is updated with the label:
```
<div aria-relevant="all" aria-live="polite" class="arialive wb-inv">
	<span class="field-name">First Name</span> <strong class="required" aria-required="true">(required)</strong> <strong id="fname1-error" class="error"><span class="label label-danger"><span class="prefix">Error&nbsp;1: </span>Please enter at least 2 characters.</span></strong>
</div>
```

The check is fired every time the jQuery Validation showErrors function is called without submission (so every keypress or field blur where there is an error). Here is the code:

```js
// Put focus on the error if the errors are generated by an attempted form submission
if ( submitted ) {

	// Assign focus to $summaryContainer
	$summaryContainer.trigger( setFocusEvent );
} else {

	// Update the aria-live region as necessary
	i = 0;
	for ( key in errorMap ) {
		if ( errorMap.hasOwnProperty( key ) ) {
			i += 1;
			break;
		}
	}
	if ( i !== 0 ) {
		len = $errors.length;
		for ( i = 0; i !== len; i += 1 ) {
			label = $errors[ i ].parentNode;
			if ( label.getAttribute( "for" ) === key ) {
				labelString = label.innerHTML;
				if ( labelString !== ariaLive.innerHTML ) {
					ariaLive.innerHTML = labelString;
				}
				break;
			}
		}
	} else if ( ariaLive.innerHTML.length !== 0 ) {
		ariaLive.innerHTML = "";
	}
}
```
Seems complicated but works pretty well.
